third_party/libunwindstack/tests/PeCoffUnwindInfoUnwinderX86_64Test.cpp

```diff
@@ -0,0 +1,1000 @@
+  std::unique_ptr<MemoryFake> process_mem_fake_;
+
+  // Stack pointer, 16-byte aligned, starting with a large value as we grow downwards
+  uint64_t stack_pointer_ = std::numeric_limits<uint64_t>::max() & ~0x0F;
+};
+
+TEST_F(PeCoffUnwindInfoUnwinderX86_64Test, eval_fails_inconsistent_num_codes) {
+  UnwindInfo unwind_info;
+  unwind_info.num_codes = 2;
+
+  RegsX86_64 regs;
+  constexpr uint64_t kCodeOffset = 0x0;
+  uint64_t frame_pointer;
+  bool frame_pointer_used;
+
+  EXPECT_FALSE(unwinder_.Eval(process_mem_fake_.get(), &regs, unwind_info, kCodeOffset,
+                              &frame_pointer, &frame_pointer_used));
+}
+
+TEST_F(PeCoffUnwindInfoUnwinderX86_64Test, eval_fails_with_unsupported_machframe_instruction) {
+  UnwindCode unwind_code;
+  unwind_code.code_and_op.code_offset = 1;
+  unwind_code.code_and_op.unwind_op_and_op_info = PackUnwindOpAndOpInfo(UWOP_PUSH_MACHFRAME, 0x00);
+
+  UnwindInfo unwind_info;
+  unwind_info.unwind_codes.emplace_back(unwind_code);
+  unwind_info.num_codes = unwind_info.unwind_codes.size();
+
+  RegsX86_64 regs;
+  constexpr uint64_t kCodeOffset = 0x0;
+  uint64_t frame_pointer;
+  bool frame_pointer_used;
+
+  EXPECT_FALSE(unwinder_.Eval(process_mem_fake_.get(), &regs, unwind_info, kCodeOffset,
+                              &frame_pointer, &frame_pointer_used));
+}
+
+TEST_F(PeCoffUnwindInfoUnwinderX86_64Test, eval_fails_with_unsupported_unwind_instruction) {
+  UnwindCode unwind_code;
+  unwind_code.code_and_op.code_offset = 1;
+  unwind_code.code_and_op.unwind_op_and_op_info =
+      PackUnwindOpAndOpInfo(static_cast<UnwindOpCode>(6), 0x00);
+
+  UnwindInfo unwind_info;
+  unwind_info.unwind_codes.emplace_back(unwind_code);
+  unwind_info.num_codes = unwind_info.unwind_codes.size();
+
+  RegsX86_64 regs;
+  constexpr uint64_t kCodeOffset = 0x0;
+  uint64_t frame_pointer;
+  bool frame_pointer_used;
+
+  EXPECT_FALSE(unwinder_.Eval(process_mem_fake_.get(), &regs, unwind_info, kCodeOffset,
+                              &frame_pointer, &frame_pointer_used));
+  EXPECT_EQ(unwinder_.GetLastError().code, ERROR_INVALID_COFF);
+}
+
+TEST_F(PeCoffUnwindInfoUnwinderX86_64Test, eval_succeeds_pushed_registers_only) {
+  std::vector<PushOp> push_ops;
+  push_ops.emplace_back(PushOp{RDI, 0x100});
+  push_ops.emplace_back(PushOp{RSI, 0x200});
+  push_ops.emplace_back(PushOp{R12, 0x300});
+
+  constexpr uint64_t kReturnAddress = 0x2000;
+
+  UnwindInfo unwind_info;
+  StackFrameOptions options;
+  options.return_address = kReturnAddress;
+  PushStackFrame(options, push_ops, {}, {}, &unwind_info);
+
+  RegsX86_64 regs;
+  // We use unwind_info.prolog_size to unwind using all unwind codes.
+  uint64_t code_offset = unwind_info.prolog_size;
+  uint64_t frame_pointer;
+  bool frame_pointer_used;
+
+  regs.set_sp(stack_pointer_);
+
+  EXPECT_TRUE(unwinder_.Eval(process_mem_fake_.get(), &regs, unwind_info, code_offset,
+                             &frame_pointer, &frame_pointer_used));
```

Comment:
Not sure if that makes sense: Should we also test for the correct value of `frame_pointer_used` in all cases when `Eval` returned `true`. I assume this would be a contract the user could rely on?

Reply:
Yes, I think that makes sense (that it's false when no frame pointer is used) and is part of the expected behavior. 

Done.